docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,164 @@
+cause connections to be opened when an LNS has lost the race against another.
+
+Several approaches were explored to understand the overall impact of
+refactoring, whether the changes to the current code base and design would be
+large or small and worth the benefits they bring. The approaches explored can
+be summed up as follows:
+
+1. Add a queue to `LoRaDevice` for all device-related operations. There is one
+   that exists today but it is used to service uplink data requests only. The
+   idea would be to extend it to encompass all other requests and operations.
+
+1. Turn the LNS implementation to be (logically) single-threaded so that not
+   only does it make it easier to reason about the code (except perhaps with
+   regards to re-entrancy), but it also enables use of very simple data
+   structures without the need for locking. This is possible because the bulk
+   of LNS is I/O bound. There is practically no CPU-intensive code that
+   executes between each await.
+
+1. Simplify the entire flow and processing of a message into a simple
+   request-response model that can be manipulated and reasoned about much more
+   easily. That is, all methods receiving and processing a `LoRaRequest`
+   return `Task<DownlinkMessage>` instead of `void` or `Task`. This allows all
+   `Task.Run` uses to be moved to a single and central point when a message is
+   received instead of being littered throughout the code base. The main
+   message loop could also be made responsible for central
+   error-handling/logging and sending of downlink messages when the processing
+   of a request has completed. This would just require regular use of tasks.
+   Next, all `Task`-based operations on `LoRaDevice` could be naturally and
+   _implicitly_ queued and then executed in a mutually exclusive manner.
+
+Another problem is that the joining of a device takes a separate execution
+path than the uplink message handling and because the two are not
+synchronized, there is room for race conditions between devices sharing the
+same `DevAddr` but using different activation methods (OTAA and ABP). To
+illustrate with an exampe, suppose an LNS has an empty device cache and a
+device named A with a `DevAddr` of 1 (from a previous join) using OTAA
+re-joins. Since the device is not in the cache, the LNS issues a request to
+the server function to search for the device on IoT Hub and obtain its primary
+key that's needed to open a device connection. Next, a `LoRaDevice` is
+_minimally initialzed_ for the purpose of loading its twin and the twin data
+requested over the device conneciton. During this time, suppose that another
+device named B sharing the same `DevAddr` as device A but using ABP activation
+sends an uplink message to the LNS. Since the `DevAddr` is not in the cache,
+the LNS will issues a search request to the server function to return all
+devices using the `DevAddr` in question.
```

Comment:
```suggestion
devices using the `DevAddr` in question eventually causing a load of the device twins for the same DevEui.
```

Reply:
Sorry, this was an older version that should have never been committed and was superseded by 60b5aaffe328f6e3b6dcfeeacdfb8ea02f0bb92b by you.